content/TypeHinting/TypeHinting.q.md

```diff
@@ -0,0 +1,1346 @@
+
+ - It prints `45.786`.
+
+ - It produces an error.
+
+ - It prints `Car Object ( [driver:protected])`.
+
+ - It prints `Person Object ( [name:protected] => 45.786 )`.
+
+ - It prints `Car Object ( [driver:protected] => Person Object ( [name:protected] => 45.786 ) )`.
+
+:::
+
+
+:::
+
+/// type=REPL
+
+```php
+<?php
+declare(strict_types = 1);
+
+class Person 
+{
+  protected $name;
+
+  public function __construct(string $name)
+  {
+    $this->name = $name;
+  }
+
+  public function setName(string $name)
+  {
+    $this->name = $name;
+  }
+
+  public function getName() : string
+  {
+    return $this->name;
+  }
+}
+
+class Car
+{
+  protected $driver;
+
+  public function __construct(Person $driver)
+  {
+    $this->driver = $driver;
+  }
+}
+
+$personObject = new Person(45.786);
+$myCar = new Car($personObject);
+print_r($myCar);
+?>
+```
+/// type=MS, answer=[4,5]
+
+Why does the statement `$personObject = new Person(45.786);` on line 34 produce an error?
+
+ - There is no semicolon `;` at the end of the statement on line 34.
+
+ - There is no `new` keyword specified before `Person(45.786)` on line 34.
+
+ - On line 34, the argument `45.786` is not enclosed in double quotes `""`.
+
+ - The `__construct()` method of the `Person` class only accepts an argument of type `string`.
+
+ - In the statement `$personObject = new Person(45.786);` on line 34, the argument `45.786` is invalid.
+
+
+/// type=SS, answer=[2]
+
+Remove the statement `$personObject = new Person(45.786);` on line 34. Execute the program. What is its output?
+
+ - It prints `45.786`.
+
+ - It produces an error.
+
+ - It prints `Car Object ( [driver:protected])`.
+
+ - It prints `Person Object ( [name:protected] => 45.786 )`.
+
+ - It prints `Car Object ( [driver:protected] => Person Object ( [name:protected] => 45.786 ) )`.
+
+:::
+
+
+:::
+
+/// type=REPL
+
+```php
+<?php
+declare(strict_types = 1);
+
+class Person 
+{
+  protected $name;
+
+  public function __construct(string $name)
+  {
+    $this->name = $name;
+  }
+
+  public function setName(string $name)
+  {
+    $this->name = $name;
+  }
+
+  public function getName() : string
+  {
+    return $this->name;
+  }
+}
+
+class Car
+{
+  protected $driver;
+
+  public function __construct(Person $driver)
+  {
+    $this->driver = $driver;
+  }
+}
+
+
+$myCar = new Car($personObject);
+print_r($myCar);
+?>
+```
+/// type=SS, answer=[2]
+
+In the satement `$myCar = new Car($personObject);` on line 35, replace `$personObject` with `"John"`. Execute the program. What is its output?
+
+ - It prints `John`.
+
+ - It produces an error.
+
+ - It prints `Car Object ( [driver:protected])`.
+
+ - It prints `Person Object ( [name:protected] => John )`.
+
+ - It prints `Car Object ( [driver:protected] => Person Object ( [name:protected] => John ) )`.
+
+:::
+
+
+:::
+
+/// type=REPL
```

Comment:
The question here does not require changes in the sample code. Are we required to add `readonly=true` here?

Reply:
Yes, because this is connected to the previous question which we allow learners to edit the code. This is to enforce learners to do their experiment with the code while trying to figure out the answer. Please check. Thank you.

Reply:
Noted, thank you.